File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Go/2c3304a41936494ed4f872146dd83ac6.go
package main

import (
	"fmt"
	"math"
	"strings"
)

type Pair struct {
	fi, se int
}

const (
	M     = 1000000007
	MAXI  = 1 << 30
	eps   = 1e-11
	id9   = 40
)

func reset(dp [][]int) {
	for i := range dp {
		for j := range dp[i] {
			dp[i][j] = 0
		}
	}
}

func main() {
	var s, t string
	fmt.Scanln(&s)
	fmt.Scanln(&t)

	sz1 := len(s)
	sz2 := len(t)

	dp := make([][]int, sz1+1)
	for i := range dp {
		dp[i] = make([]int, sz2+1)
	}

	reset(dp)

	for i := 0; i < sz1; i++ {
		for j := 0; j < sz2; j++ {
			dp[i+1][j+1] = (dp[i+1][j] + (boolToInt(s[i] == t[j]) * (dp[i][j] + 1) % M)) % M) % M
		}
	}

	re := 0
	for i := 0; i < sz1; i++ {
		re = (re + dp[i+1][sz2]) % M
	}
	fmt.Println(re)
}

func boolToInt(b bool) int {
	if b {
		return 1
	}
	return 0
}

func dist(a, b Pair) float64 {
	return math.Sqrt(sqr(float64(a.fi)-float64(b.fi)) + sqr(float64(a.se)-float64(b.se)))
}

func sqr(x float64) float64 {
	return x * x
}

func cross(c, a, b Pair) int {
	return (a.fi-c.fi)*(b.se-c.se) - (b.fi-c.fi)*(a.se-c.se)
}

func crossOper(c, a, b Pair) int {
	t := (float64(a.fi)-float64(c.fi))*(float64(b.se)-float64(c.se)) - (float64(b.fi)-float64(c.fi))*(float64(a.se)-float64(c.se))
	if math.Abs(t) <= eps {
		return 0
	}
	if t < 0 {
		return -1
	}
	return 1
}

func isIntersect(a, b, c, d Pair) bool {
	return crossOper(a, b, c)*crossOper(a, b, d) < 0 && crossOper(c, d, a)*crossOper(c, d, b) < 0
}

func isMiddle(s, m, t float64) bool {
	return math.Abs(s-m) <= eps || math.Abs(t-m) <= eps || (s < m) != (t < m)
}

func isLetter(c byte) bool {
	return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z')
}

func isDigit(c byte) bool {
	return c >= '0' && c <= '9'
}

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Python/2c3304a41936494ed4f872146dd83ac6.py
import math

def dist(a, b):
    return math.sqrt((a[0] - b[0])**2 + (a[1] - b[1])**2)

def distR(a, b):
    return (a[0] - b[0])**2 + (a[1] - b[1])**2

def cross(c, a, b):
    return (a[0] - c[0])*(b[1] - c[1]) - (b[0] - c[0])*(a[1] - c[1])

def crossOper(c, a, b):
    t = (a[0] - c[0])*(b[1] - c[1]) - (b[0] - c[0])*(a[1] - c[1])
    if abs(t) <= 1e-11:
        return 0
    return -1 if t < 0 else 1

def isIntersect(a, b, c, d):
    return crossOper(a, b, c) * crossOper(a, b, d) < 0 and crossOper(c, d, a) * crossOper(c, d, b) < 0

def isMiddle(s, m, t):
    return abs(s - m) <= 1e-11 or abs(t - m) <= 1e-11 or (s < m) != (t < m)

def isUpperCase(c):
    return c >= 'A' and c <= 'Z'

def isLowerCase(c):
    return c >= 'a' and c <= 'z'

def isLetter(c):
    return isUpperCase(c) or isLowerCase(c)

def isDigit(c):
    return c >= '0' and c <= '9'

def toLowerCase(c):
    return chr(ord(c) + 32) if isUpperCase(c) else c

def toUpperCase(c):
    return chr(ord(c) - 32) if isLowerCase(c) else c

def toString(n):
    return str(n)

def toInt(s):
    return int(s)

def toInt64(s):
    return int(s)

def toDouble(s):
    return float(s)

def stoa(s):
    return list(map(int, s.split()))

def atos(a):
    return ' '.join(map(str, a))

def atov(a):
    return a

def vtoa(a):
    return a

def stov(s):
    return list(map(int, s.split()))

def vtos(a):
    return ' '.join(map(str, a))

class Fraction:
    def __init__(self, a=0, b=1):
        d = math.gcd(a, b)
        a //= d
        b //= d
        if b < 0:
            a = -a
            b = -b
        self.a = a
        self.b = b

    def __str__(self):
        return str(self.a) + "/" + str(self.b)

    def __add__(self, other):
        return Fraction(self.a * other.b + other.a * self.b, self.b * other.b)

    def __sub__(self, other):
        return Fraction(self.a * other.b - other.a * self.b, self.b * other.b)

    def __mul__(self, other):
        return Fraction(self.a * other.a, self.b * other.b)

    def __truediv__(self, other):
        return Fraction(self.a * other.b, self.b * other.a)

dp = [[0] * 5005 for _ in range(5005)]

s = input()
t = input()
sz1 = len(s)
sz2 = len(t)
for i in range(sz1):
    for j in range(sz2):
        dp[i + 1][j + 1] = (dp[i + 1][j] + (s[i] == t[j]) * (dp[i][j] + 1)) % (10**9 + 7)
re = 0
for i in range(sz1):
    re = (re + dp[i + 1][sz2]) % (10**9 + 7)
print(re)
